File: table_creator.pl
# April 2019
# Julia Raices
# Program to take output from blat with exons and intronsdo everything I ever did with this tables
#!/usr/bin/perl

use strict; # doesn't let you use variables without declaring them

# declared variables:
my ($i, $k, $n, $j, $l, $trans, $outname, $line, $key, $thing); # counters and strings
my (@reads, @transcript, @sorted, @exons, @temp, @temp2, @temp3, @genez, @list, @f); # arrays with raw data
my (%values); # hashs with printable data

# undefine variables, so that they don't - by chance - start not empty
undef $i;
undef $j;
undef $l;
undef $k;
undef $n;
undef @sorted;
undef @genez;
undef $thing;
undef $trans;
undef $line;
undef $key;
undef $outname;
undef @reads;
undef @list;
undef @f;
undef @transcript;
undef @exons;
undef @temp;
undef @temp2;
undef @temp3;
undef %values;

$j = 0;

# sees what is the argument (after the program was called) and stores it. It should be the name of file to be used by the program
for($i=0; $i<=$#ARGV; $i+=2){
	if($ARGV[$i] eq "-trans"){
		$trans=$ARGV[$i+1];
	}
}
for($i=0; $i<=$#ARGV; $i+=2){
	if($ARGV[$i] eq "-out"){
		$outname=$ARGV[$i+1];
	}
}

# check if files have been designated correctly, if they are not, it tells you how to use the program and exits
if($trans eq "" | $outname eq ""){
	print STDERR "Program usage: perl table_creator.pl -trans BLAT_OUTPUT_TRANSCRIPTS -out OUTPUT_NAME\n\t-trans: Table with exons and introns that compose transcripts (blat output).\n\t-out: Name and location of the output file.\n";
	exit 0;
}

# open log file and adds data from this time program was run
open (LOG, ">>table.log");

print LOG "\n".(localtime)."\nTable for transcripts data: $trans\nOutput file: $outname\n";

# finds if input files actually exists and are openable, if not, prints error in log and in stderr and exists program
#unless(open(TRA, $trans) || open(OUTPUT, '>', $outname)){
#	print STDERR "Couldn't open file, please check if file exist and if you have permission to read it.\n";
#	print LOG "Error opening file.\n";
#	exit 0;
#}

system("awk '{print $10}' $trans | sort | uniq >genelist.tmp");

open(GEN, "genelist.tmp");
while(<GEN>){
	$line=$_;
	chomp $line;
	if($line=~/Symbol/){}
	else{
		$genez[$#genez+1]=$line;
		#print $line, '\n';
	}
}
close(GEN);
system("yes | rm genelist.tmp");
undef $line;

open(F,$trans);
@list=<F>;
close F;

#foreach $thing (@genez){
#	@f=grep /$thing/, @list;

	#open(TMP, '>', $thing);
	#print TMP @f;
	#@sorted = sort {(split(/\t/, $b))[11] <=> (split(/\t/, $a))[11]} @f;
	#	@sorted = sort { my @a = split /\t/, $a; my @b = split /\t/, $b; $a[11] <=> $b[11]; } @f;

	#foreach $line (@sorted){
	#print $line;
		#AQUI
		#}
		#}
undef $line;


open(OUTPUT, '>', $outname);

# open output files, and prints header in it
print OUTPUT "gene\ttranscript\texon_intron\tuniqID\n";

open(TRA, $trans);
# get transcript data from table
while(<TRA>){ # as far as there are things/lines in the file
	$line=$_; # read it line by line
	chomp $line; # remove end of line character
	@reads=split(/\t/, $line); # split line where there's a tab
	if($line=~/Symbol/){} # if line is the header, ignore it
	@transcript=split(/\//,$reads[9]); # for the 10th column (perl starts counting at 0) split it in th "/" - this is not necessary, but it will make it prettier
	if(index($reads[13], 'intron') != -1){#here we check if the match was for an intron, and if it was, we will do things a little differently, because introns have a different form of annotation
		# the annotation for introns goes as: intron_FBgn###:#_FBgn###:#
		@temp=split(/_/,$reads[13]); # this separetes the beggining of the intron from its end
		@temp2=split(/:/, $temp[1]); # this stores the 1st exon that borders the intron
		@temp3=split(/:/, $temp[2]); # and this, is the 2nd exon
		if($temp2[0]!=$temp3[0]){#to make sure that introns are contained within the name gene
			print LOG "Error: there's an intron between genes $temp2[0] and $temp3[0]\n";
			# if the intron is between different genes (what shouldn't happen) mark it on the log!
			$exons[0]="$temp2[0]_$temp3[0]";
			$exons[1]="$temp2[1]_$temp3[1]";
			$j++;
		}
		else{
			$exons[0]="$temp2[0]";
			$exons[1]="$temp2[1]_$temp3[1]";
			$l++;
		}
	}
	else{
		@exons=split(/:/,$reads[13]);
	}

	if(exists($values{$transcript[1]})){
		#if this transcript has already been found, add to it's entry
		$values{$transcript[1]}[2]="$values{$transcript[1]}[2],$exons[1]";
		if($values{$transcript[1]}[0]==$exons[0]){}
		else{
			$values{$transcript[1]}[0]="$values{$transcript[1]}[0],$exons[0]";
			print LOG "ERROR: transcript $transcript[1] has more than one reference gene!\n";
			#mark in the log if the transcript is matched to differnt genes
		}
	}
	else{#otherwise create a new entry
		$values{$transcript[1]}[0]="$exons[0]";
		$values{$transcript[1]}[1]="$transcript[1]";
		$values{$transcript[1]}[2]="$exons[1]";
	}
	$k++;

}

## prints data in output. gives an NA (Not Avaiable) for not avaiable dn/ds data.
foreach $key (keys %values){ # as a foreach was used, there are only uniq copies of each CG in the final output file.
	$n++;
	print OUTPUT "$values{$key}[0]\t$values{$key}[1]\t$values{$key}[2]\t$values{$key}[0]/$values{$key}[1]_$values{$key}[2]\n";
}

# print what was done and found in log file
print LOG "\tThere were $k lines in the Transcript Table and $n of them were printed in the output.\n\t\t$j lines had weird introns (belonging to two different genes).\n\t\t$l lines corresponded to normal introns.\n";

# close files used and exit program.
close LOG;
close OUTPUT;
close TRA;

exit;

File: table_creator.jl
# Julia Raices
# November 2019
# Program to create tables with exon/intron data from blat output

using CSV
using DataFrames
using Dates

given1=""
given2=""

given1 = ARGS[1] #reads the first argument after program call
#given1 = "test.out"
#given1="dmel_analysis/whole_body/isoseq/both_sexes_mixed/transcripts/both_exonintron.blat.final"
given2 = ARGS[2] #reads the 2nd argument after the program call
#given2 = "testandoJL.out"

if isempty(given1) | isempty(given2) | (given1==given2)
	print("ERROR: the given files are not usable\n")
	exit()
end

trans = CSV.read(given1, delim="\t", header=true) # opens the first argument as a tsv table
outputz=open(given2, "w") #opens the 2nd argument to write on it (it's the output
log=open("maketablesJL.log", "w") #also opens a log so that any errors os stuff can be recorded
#oooo=open("meow.test", "w")

print(outputz, "#Flags meaning: 1 - Intron between genes\n\t# 2 - transcript has parts starting or ending inside one another\n\t# 3 - transcript has more than one reference\n\t# 4 - transcript error (different transcripts treated as the same\ngene\ttranscript\texon_intron\tuniqID\tflags\n")
# first prints header and information on the output
print(log, "\n\n", Dates.now(),"\n") #writes the date and time on the log

unique_trans=unique(trans[:10]) #selects the 10th column of the blat file and gets all the names that appear in it, creating a new list with those unique names
#length(unique_trans) #amount of unique reads/transcripts
# set the counters to 0
toout=0
intrs=0
Wintrs=0
Nintrs=0
exns=0
# make a dataframe to store the data for the transcripts
list=DataFrame(transc=String[], times=Int64[], gene=String[], exons=String[], flags=String[])

# for every transcript in the unique list
for k in 1:length(unique_trans)
	temp=trans[occursin.(unique_trans[k], trans[:10]), :] # create a temporary file with all instances of that transcript, and only that transcript
	temp2=sort!(sort!(temp, :13), :12) # sort the temporary file accordinf to the end and beggining of the transcript (in the reference gene)
	#set future variables to null
	exon=""
	temp2.transcripts=repeat("",length(temp2[:,:1]))
	gene_toprint = ""
	transc_toprint = ""
	flagging=""
	q=0
	#for every match in our sorted temporary file
	for i in 1:length(temp2[:,:1])
		if occursin("intron", temp2[i,:14])
		# if the match is to an intron, split it and make sure to only have the exons between which the intron is to put on  the final table
			meh, intronPA, intronPB = split(temp2[i,:14], "_")
			genePA, exonPA=split(intronPA, ":")
			genePB, exonPB=split(intronPB, ":")
			if (genePA != genePB)
				#if the intron is between different genes, mark it as an error in the log, and add a flag to the output
				print(log, "ERROR: there's an intron between genes ", genePA, " and ", genePB, "\n")
				if flagging==""
					#if there's no flag, add this one only
					flagging='1'
				else
					#if there's a flag, add this one with a separator between it and the previous flag
					flagging=flagging*'_'*'1'
				end
				temp2[i,:22] = genePA*'_'*genePB
				#since the intron is between two genes, add both to the gene parameter
				temp2[i,:14] = exonPA*'_'*exonPB
				# add the number of the exons between which the intron is
				global Wintrs=Wintrs+1 #increase counter
			else
				temp2[i,:22] = genePA
				#since the intron is within a gene, add it to the gene parameter
				temp2[i,:14] = exonPA*'_'*exonPB
				# add number of exons between which the intron is
				global Nintrs=Nintrs+1 #increase counter
			end
			global intrs=intrs+1 #increase counter
		else
			#if the match is not to an intron, is to an exon, and then we just add the gene and exon to its parameters
			temp2[i,:22], temp2[i,:14] = split(temp2[i,:14], ":")
			global exns=exns+1
		end
		#make new temporary file with everyline except the one we are looking at (so we can compare the one we are looking at to the others)
		temp3 = temp2[setdiff(1:end, i), :]
		for j in 1:length(temp3[:1])
			# for every line in the new temporary file
			if temp2[i, :12] in collect(temp3[j,:12]:temp3[j,:13]) || temp2[i, :13] in collect(temp3[j,:12]:temp3[j,:13])
				# if the start or the end of the match we are looking at is between the start and end of the line we are comparing it to print an error to the log, and flag the read
				print(log, "ERROR: transcript ", unique_trans[k], " has parts starting or ending inside one another.\n")
				if flagging==""
					#if there's no flag, add this one only
					flagging='2'
				else
					#if there's a flag, add this one with a separator between it and the previous flag
					flagging=flagging*'_'*'2'
				end
			end
			if temp2[i, :22] != temp3[j, :22]
				# if the gene is not the same between two of the matches flag it and add an error note to the log
				print(log, "ERROR: transcript ", unique_trans[k], " has more than one reference gene (", temp2[i, :22], " and ", temp3[j, :22], " )!\n")
				if flagging==""
					#if there's no flag, add this one only
					flagging="3"
				else
					#if there's a flag, add this one with a separator between it and the previous flag
					flagging=flagging*'_'*"3"
				end
				q=q+1
			end
		end

		if exon==""
			#if this is the first exon to be added to this transcript, add it to the variable
			exon=temp2[i,:14]
		else
			# otherwise add it with a comma between the previous and the new numbers
			exon=(exon*','*temp2[i, :14])
		end

		if gene_toprint==""
			#if it's the first time we are giving the transcript a reference gene, just do it
			gene_toprint = temp2[i, :22]
		elseif gene_toprint!=temp2[i,:22]
			# if we have a reference gene stablished, but this match has a different one, we add it to the previous with a spacer (in this case a pipe), and add an error to the log, and flag the transcript
			print(log, "ERROR: transcript ", unique_trans[k], " has more than one reference gene (", temp2[i, :22], " and ", gene_toprint, " )!\n")
			if flagging==""
				#if there's no flag, add this one only
				flagging="3"
			else
			#if there's a flag, add this one with a separator between it and the previous flag
				flagging=flagging*'_'*"3"
			end
			gene_toprint=gene_toprint*'|'*temp2[i,:22]
		else
			# if we have a reference gene stored and it's the same as the one now, just keep it, bc everything is fine
		end

		if transc_toprint==""
			# check if the transcript was in fact the same
			# if it is, everything is fine, if it's empty, make it that transcript
			transc_toprint = temp2[i, :10]
		elseif transc_toprint != temp2[i,:10]
			# if it's different, make an error note and add a flag
			transc_toprint = transc_toprint*"|"*temp2[i,:10]
			print(log, "ERROR: more than one transcript being analysed together! (", transc_toprint, " and ", temp2[i, :10], ")\n")
			if flagging==""
				#if there's no flag, add this one only
				flagging="4"
			else
			#if there's a flag, add this one with a separator between it and the previous flag
				flagging=flagging*'_'*"4"
			end
		end
	end
	# add this transcript to the dataframe we made for it:	
	push!(list, (gene=gene_toprint, transc=transc_toprint, exons=exon, times=length(temp2[:1]), flags=flagging))
	global toout=toout+1 # increase counter
end

#toout
#intrs
#Wintrs
#Nintrs
#exns

#if length(unique_trans)==length(list[:,1])
#	print("YES!!!")
#else
#	print("FUUUUUCK!!!!!!")
#end

#length(unique_trans)
#length(list[:,1])

print(outputz, list)

print(log, "\tThere were ", length(trans[:10]), " lines in the Transcript Table and ", toout, " of them were printed in the output. \n\t\tFrom ", intrs, " introns, ", Wintrs, " were weird introns (belonging to two different genes) and ", Nintrs, " lines corresponded to normal introns.\n")

close(outputz)
close(log)


